### PR TITLE
Var derivation funcs 133716643

### PR DESCRIPTION
**Pull request by jjdelc**

### CODE REVIEW

File: source/includes/_feature-deriving.md [hunk 1]
```diff
@@ -395,4 +395,158 @@ POST'ing to the private variables catalog a Shoji Entity containing a ZCL functi
...
+Receives a numeric column and returns a categorical column where each
+category represents a bin of the numeric values.
+Each category is annotated with a "boundaries" member that contains the 
+lower/upper bound of each bin.
+
+```json
+{
+    "function": "bin",
+    "args": [
+        {"variable": "http://.."}
```
Comment (nealrichardson) [inline source/includes/_feature-deriving.md]: This is not a complete example.

File: source/includes/_feature-deriving.md [hunk 2]
```diff
@@ -395,4 +395,158 @@ POST'ing to the private variables catalog a Shoji Entity containing a ZCL functi
...
+    ]
+}
+```
+
+#### case
+Returns a categorical column with its categories following the specified
+conditions from different columns on the dataset. [View Case Statements](#Case-statements)
+
+#### cast
```
Comment (nealrichardson) [inline source/includes/_feature-deriving.md:460]: The numeric cast is the only useful one here for doing calculations on categoricals using their numeric values on the fly. I guess you could use this to derive a variable. The categorical implementation seems really fragile for that. I'd discourage that use, and link also to the variables/id/cast/ resource for changing the type of a variable.

File: source/includes/_feature-deriving.md [hunk 3]
```diff
@@ -395,4 +395,158 @@ POST'ing to the private variables catalog a Shoji Entity containing a ZCL functi
...
+Returns a new column with its type and values casted. Not applicable
+on arrays or date columns.
```
Comment (nealrichardson) [inline source/includes/_feature-deriving.md]: Can't you format dates as text somehow? Different function? (Just curious, not relevant for them.)

File: source/includes/_feature-deriving.md [hunk 4]
```diff
@@ -395,4 +395,158 @@ POST'ing to the private variables catalog a Shoji Entity containing a ZCL functi
...
+
+```json
+{
+    "function": "cast",
+    "args": [
+        {"variable": "http://.."},
```
Comment (nealrichardson) [inline source/includes/_feature-deriving.md]: Please use complete examples, not ellipsized URLs.

File: source/includes/_feature-deriving.md [hunk 5]
```diff
@@ -395,4 +395,158 @@ POST'ing to the private variables catalog a Shoji Entity containing a ZCL functi
...
+For categorical types it is necessary to indicate the categories as a type
+definition.
+
+```json
+{
+    "function": "cast",
+    "args": [
+        {"variable": "http://.."},
+        {"value": {
+                "class": "categorical",
```
Comment (nealrichardson) [inline source/includes/_feature-deriving.md:494]: If you're going to give this example, you'll have to note that the "value" is not a string but a type object. But I wonder whether you need to include this case.

File: source/includes/_feature-deriving.md [hunk 6]
```diff
@@ -395,4 +395,158 @@ POST'ing to the private variables catalog a Shoji Entity containing a ZCL functi
...
+#### get
+Returns a subvariable from an array as a categorical column.
+The first argument should be the URL of an array variable and the 
+second attribute a `value` containing the string ID of the desired
+subvariable.
+
+
+```json
+{
+    "function": "get",
```
Comment (nealrichardson) [inline source/includes/_feature-deriving.md]: Do you have a test showing this function used? Never seen it in the wild.

File: source/includes/_feature-deriving.md [hunk 7]
```diff
@@ -395,4 +395,158 @@ POST'ing to the private variables catalog a Shoji Entity containing a ZCL functi
...
+    "args": [
+        {"variable": "http://.."},
+        {"value": "Text missing reason"},
+        {"value": <int_code_to_use>}
+    ]
+}
+```
+
+#### normalize
+Receives a numeric value and returns a numeric column with
```
Comment (nealrichardson) [inline source/includes/_feature-deriving.md]: Example? Really, this exists?

File: source/includes/_feature-deriving.md [hunk 8]
```diff
@@ -395,4 +395,158 @@ POST'ing to the private variables catalog a Shoji Entity containing a ZCL functi
...
+normalized values such that the its sum == its length
+
+#### row
+Returns a numeric column with row indices.
```
Comment (nealrichardson) [inline source/includes/_feature-deriving.md]: Base 0?

File: source/includes/_feature-deriving.md [hunk 9]
```diff
@@ -395,4 +395,158 @@ POST'ing to the private variables catalog a Shoji Entity containing a ZCL functi
...
+    "args": [
+        {"variable": "http://.."},
+        {"value": "subvariable id"}
+    ]
+}
+```
+
+
+#### missing
+Returns a column marked as missing accorging to a value. This is generally
```
Comment (nealrichardson) [inline source/includes/_feature-deriving.md]: Where is this used? Never seen it.

File: source/includes/_feature-deriving.md [hunk 10]
```diff
@@ -395,4 +395,158 @@ POST'ing to the private variables catalog a Shoji Entity containing a ZCL functi
...
+Returns a numeric column with row indices.
+
+```json
+{
+    "function": "row",
+    "args": []
+}
+```
+
+#### unmissing
```
Comment (nealrichardson) [inline source/includes/_feature-deriving.md]: Likewise have never seen this and question the value of exposing/documenting it.

Comment (nealrichardson) [reply source/includes/_feature-deriving.md]: Also might need an example of the output for this and "missing", if you're including them.

File: source/includes/_feature-deriving.md (hunk 1 above)
Comment (nealrichardson) [reply source/includes/_feature-deriving.md]: Can you specify the number of bins? Show an example of the output too.

File: source/includes/_feature-deriving.md [hunk 11]
```diff
@@ -395,4 +395,158 @@ POST'ing to the private variables catalog a Shoji Entity containing a ZCL functi
 
 ### Other transformations
```
Comment (nealrichardson) [inline source/includes/_feature-deriving.md:397]: How about some examples of the basic +, -, ==, <, etc. functions?

Comment (jjdelc) [reply source/includes/_feature-deriving.md:397]: I did the functions that return columns, those are functions that do not return columns but masks. We had a ticket to make columns out of boolean ops but we don't have them yet. They cannot be used to create new variables.

Comment (nealrichardson) [reply source/includes/_feature-deriving.md:397]: Fair enough about the boolean ones for now, but `{"function": "+", "args":[ {"variable":id}, {"value": 5}]}` returns a column, doesn't it?

File: source/includes/_feature-deriving.md (hunk 1 above)
Comment (jjdelc) [reply source/includes/_feature-deriving.md]: Added an example with gap size. There is no example output other than the described categorical variable.

File: source/includes/_feature-deriving.md (hunk 5 above)
Comment (jjdelc) [reply source/includes/_feature-deriving.md:494]: I think it is necessary as a way to go back and forth? I'd gladly remove it but seemed incomplete to not include one very evident type. 

I am rephrasing the line above the example.

File: source/includes/_feature-deriving.md (hunk 6 above)
Comment (jjdelc) [reply source/includes/_feature-deriving.md]: Removing it, we use it internally.

File: source/includes/_feature-deriving.md (hunk 9 above)
Comment (jjdelc) [reply source/includes/_feature-deriving.md]: Never directly, we use it to mark variables as missing updating a filtered column with missing values, always with a filter applied, else the full column will just contain all missing values. Maybe this one isn't necessary.

The following `unmissing` function returns a column with all missing values cleared, so it returns the full raw data again. That seems more useful.

File: source/includes/_feature-deriving.md (hunk 7 above)
Comment (jjdelc) [reply source/includes/_feature-deriving.md]: Yeah, never used it. I had to read the code to figure what it does. Will remove.

File: source/includes/_feature-deriving.md (hunk 8 above)
Comment (jjdelc) [reply source/includes/_feature-deriving.md]: yes, noted

**Pull request merged**